internal/statemachine/classic_states.go

```diff
@@ -245,6 +245,20 @@ func (stateMachine *StateMachine) calculateStates() error {
 	if classicStateMachine.ImageDef.Rootfs.Tarball != nil {
 		rootfsCreationStates = append(rootfsCreationStates,
 			stateFunc{"extract_rootfs_tar", (*StateMachine).extractRootfsTar})
+		// if there are extra snaps or packages to install, these will have
+		// to be done as separate steps
+		if len(classicStateMachine.ImageDef.Customization.ExtraPPAs) > 0 {
```

Comment:
Does this need to check `classicStateMachine.ImageDef.Customization != nil` like the code below does?

Reply:
For a similar problem with determining which artifacts needed to run the `make_disk` state we added a struct tag that we could use to determine this. I'll take a look at doing something similar to reduce this duplication.

Reply:
@mwhudson the commit I just pushed reduces duplication by adding states based on struct tags. It currently only does this for image builds with a prebuilt rootfs. For image builds that are building a rootfs the only extra step that might be needed is `add_extra_ppas`. I will think about if it would be best to key off that struct tag for these image build types too. If you are happy with this approach I will fix the test coverage and write some unit tests for the function that checks the struct tags.

Reply:
I think it looks nice but maybe we should get @sil2100 to take a look too?

Reply:
I like the idea, it sounds and looks similar to what we were trying out for making the state definitions be able to map yaml->state_function, but which was dropped as there was too many conditional handling that needed to happen.
My main concern here would be the confusion factor:
1) the parameter is called `extra_step`, which doesn't quite tell you what for? Someone wanting to add some new states might think this is unconditionally executed everytime, and not just in the case of re-using a rootfs tarball. So maybe looking at the naming?
2) if I recall correctly, this is the only place where we explicitly put a state function name in, instead of relying on that happening in the state machine creation function. This feels a bit inconsistent, and feels like a bigger 'implementation design change'.

I need to think about this for a moment and will re-iterate on those when I submit my formal review.